src/git.md

```diff
@@ -0,0 +1,181 @@
+instead made to the new version of master.
+
+During rebasing, you should expect to encounter at least one "rebase conflict."
+This happens when git's attempt to reapply the changes onto the more recent
+version of master fails because your changes conflicted with other changes that
+have been made since then. You can tell that this happened because you'll see
+lines in the output that look like
+
+```
+CONFLICT (content): Merge conflict in file.rs
+```
+
+When you open these files, you'll see sections of the form
+
+```
+<<<<<<< HEAD
+Original code
+=======
+Your code
+>>>>>>> 8fbf656... Commit fixes 12345
+```
+
+This represents the lines in the file that git could not figure out how to
+rebase. The section between `<<<<<<< HEAD` and `=======` has the code from
+master, while the other side has your version of the code. You'll need to
+decide how to deal with the conflict. You may want to keep your changes,
+keep the changes on master, or combine the two.
+ 
+Generally, resovling the conflict consists of two steps: First, fix the
+particular conflict. Edit the file to make the changes you want and remove the
+`<<<<<<<`, `=======` and `>>>>>>>` lines in the process. Second, check the
+surrounding code. If there was a conflict, its because someone else changed the
+same code you did. That means its likely there are some logical errors lying
+around too!
+
+Once you're all done fixing the conflicts, you need to stage the files that had
+conflicts in them via `git add`. Afterwards, run `git rebase --continue` to let
+git know that you've resolved the conflicts and it should finish the rebase.
+Finally, once the rebase has succeeded, you'll want to update the associated
+branch on your fork with `git push -f`.
+
+## Advanced Rebasing
+
+Sometimes, you may want to perform a more complicated rebase. There are two
+common scenarios that might call for this.
+
+If your branch contains multiple consecutive rewrites of the same code, or if
+the rebase conflicts are extremely severe, it is possible that just trying to
+reapply the changes you made on top of the updated code will be too much of a
+headache. In this case, you can use the interactive rebase feature via
+`git rebase -i master` to gain more control over the process. This allows you
+to choose to skip commits because they represent changes you no longer need,
+edit the commits that you do not skip, or change the order in which they are
+applied.
+
+The other common scenario is if you are asked to or want to "squash" multiple
+commits into each other. The most common example of this is likely if you
+forgot to run `x.py tidy` before committing. Your PR will need to be revised,
+but a single commit at the end with message "fixup tidy issue" is usually
+unhelpful, and it is easier for everyone else if you combine that commit with
+another that has a more meaningful commit message. In this case, you'll want to
+run `git rebase -i HEAD~2` to edit the last two commits and merge them
```

Comment:
This works for the most recent two commits, but I normally rebase the whole branch while I'm at it.
```suggestion
run `git rebase -i $(git merge-base HEAD master)` to edit your commits and merge them
```
Maybe that's too much for a beginner guide though?

Reply:
I have not addressed this yet, but I'm also not quite sure how to. A slightly more general approach to this section could indeed be useful, but I don't know how to write it in a way that doesn't sacrifice clarity.

Reply:
Let's hold off on the more general approach for now, then. I'm planning to add a more 'advanced' section with things like `range-diff` anyway, I think this would be a good fit there.